A draft for the approximate multi-qubit toffoli gate algorithm

chenyizhang2000 opened the pull request.

chenyizhang2000 committed:
a draft for approximating multi-qubit toffoli gate

chenyizhang2000 committed:
Merge branch 'main' into multi_qubit_toffoli_draft

chenyizhang2000 committed:
notation_changes

chenyizhang2000 committed:
Update multi_qubit_toffoli notebook

chenyizhang2000 commented:
The new version has the following updates:
1. A draft on "concepts" containing a narrative description of the construction
2. I moved the file into qualtran.bloqs.mcmt.
3. I added some comments inside the code to meet the docstring conventions, especially for the SamplePreparationBloq. I also adopted bb.alloc(n), bb.add(OneState()), or bb.add(ZeroState()) to simpify things.
4. I modified the bloqs to make them have the same external interface as MultiAnd.

mpharrigan commented on qualtran/bloqs/mcmt/multi_qubit_toffoli.ipynb:
```diff
@@ -0,0 +1,2992 @@
+   "id": "6186bdee-982c-4061-88dc-d8ba9b62dccc",
+   "metadata": {},
+   "source": [
+    "This note presents an approximate $n$-qubit AND/Toffoli gate construction introduced in [GKZ25]. The idea is that, instead of directly computing the full AND, we randomly sample a small number of subsets of $[n]$, compute their parities, and combine those parity values through a smaller exact AND together with Clifford processing.\n",
```
Please add a title to the notebook (you can also add a little byline with your name/email as author if you'd like)

In markdown, we write

```markdown
# This is my title

This is the body text.
```

chenyizhang2000 replied:
added

mpharrigan commented on qualtran/bloqs/mcmt/multi_qubit_toffoli.ipynb:
```diff
@@ -0,0 +1,2992 @@
+    "scrolled": true
+   },
+   "outputs": [
+    {
```
In jupyterlab or jupyter notebook, use "Edit > Clear output of all cells" before committing with git. We store un-executed versions of the notebooks in the repository or spurious changes to the outputs would pollute the git history

chenyizhang2000 replied:
got it - I've cleared it

mpharrigan reviewed (commented):
This is fantastic! I think we should name the file `approx_multi_toffoli.ipynb` or `approx_multi_qubit_toffoli.ipynb`

The bloq definitions are fully formed and should ultimately get "promoted" to live in a python file in `qualtran.bloqs.mcmt`. This is a bit more work, so I'm happy to merge the notebook as-is and leave the rest as a follow up; or do it now as part of the PR. The steps would be

 - take bloq class definitions and put into `approx_multi_toffoli.py`.. import these bloq classes in the concepts notebook
 - Write one or two `@bloq_example` example instantiations https://qualtran.readthedocs.io/en/latest/Autodoc.html#demo-bloqs-with-bloqexample
 - Add a file for unit tests: `approx_multi_toffoli_test.py`. Use the bloq examples to run basic validity checks
 - add en entry to qualtran_dev_tools/notebookspecs.py to make an autogenerated reference-api notebook. This can stay separate from the "concepts" notebook you've written. 

let me know if you'd like to do that or merge as-is

chenyizhang2000 committed:
Add a .py file for the bloq definitions

chenyizhang2000 commented:
Moving the bloq definitions into a python file sounds good! I've made the following updates:

- Renamed the concepts notebook to `approx_multi_toffoli.ipynb`.
- Moved the bloq implementations into `approx_multi_toffoli.py`.
- Updated the concepts notebook to import the bloqs from `qualtran.bloqs.mcmt`.
- Added a `NotebookSpecV2` entry for the autogenerated reference API notebook.
- Added the file `approx_multi_toffoli_file.py` for unit test, and tried to run some basic validity checks. I don't fully understand the test log though

mpharrigan commented on qualtran/bloqs/mcmt/approx_multi_toffoli_test.py:
```diff
@@ -0,0 +1,41 @@
+
+    _, _, target = bloq.call_classically(ctrl=int_to_bits(x_int, m))
+
+    assert target == int(x_int == all_ones)
```
Please add an additional small test that the notebook executes without error. At the top of the file:

```python
import qualtran.testing as qlt_testing
```

and here at the bottom

```python
@pytest.mark.notebook
def test_notebook():
    qlt_testing.execute_notebook('approx_multi_toffoli')
```

chenyizhang2000 replied:
test added

mpharrigan commented:
> I don't fully understand the test log though

Can you say more about this? The unit tests pass on my machine and on the automated CI checks on this PR. You can run the unit tests with `pytest qualtran/bloqs/mcmt/approx_multi_toffoli_test.py` specifically or `check/pytest-quick` to run all the tests for the whole project

mpharrigan commented on dev_tools/qualtran_dev_tools/notebook_specs.py:
```diff
@@ -261,6 +262,17 @@
+            qualtran.bloqs.mcmt.approx_multi_toffoli._PARITY_MASK_DOC,
+            qualtran.bloqs.mcmt.approx_multi_toffoli._APPROX_MULTI_TOFFOLI_DOC,
+        ],
+    ),
```
Please run 

```bash
python dev_tools/autogenerate-bloqs-notebooks-v2.py
```

which will create a new notebook "approx_multi_toffoli_reference". You'll need to commit that to git (via `git add`). It'll also edits docs/bloqs/index.rst to include the new notebook in the table-of-contents. you'll need to commit that too

chenyizhang2000 replied:
New notebook generated. The unit test by running `pytest qualtran/bloqs/mcmt/approx_multi_toffoli_test.py `now works fine. There might be a python package missing previously

mpharrigan commented:
We run the mypy type-checker to try to detect type errors before runtime. You can run this locally with `check/mypy`. There are a couple of minor issues. You can fix them with these changes:


```patch
diff --git a/qualtran/bloqs/mcmt/approx_multi_toffoli.py b/qualtran/bloqs/mcmt/approx_multi_toffoli.py
index 6d9e8620..946c7478 100644
--- a/qualtran/bloqs/mcmt/approx_multi_toffoli.py
+++ b/qualtran/bloqs/mcmt/approx_multi_toffoli.py
@@ -23,6 +23,8 @@ from qualtran.drawing import (
     directional_text_box,
     show_bloq,
 )
+from qualtran.simulation.classical_sim import ClassicalValRetT
+
 
 @dataclass(frozen=True)
 class MultiAndLogDepth(Bloq):
@@ -190,7 +192,7 @@ class MultiAndLogDepth(Bloq):
         junk = [left_out] + left_junk + [right_out] + right_junk
         return junk, out
 
-    def on_classical_vals(self, ctrl: np.ndarray) -> Dict[str, np.ndarray]:
+    def on_classical_vals(self, ctrl: np.ndarray) -> Dict[str, ClassicalValRetT]:
         ctrl = np.asarray(ctrl, dtype=np.uint8)
         effective_ctrl = np.equal(ctrl, np.asarray(self.concrete_cvs)).astype(np.uint8)
         junk, target = self._classical_tree(list(effective_ctrl))
@@ -226,7 +228,7 @@ class MultiAndLogDepth(Bloq):
         return f"MultiAndLogDepth(n={self.n_ctrls})"
 
     def build_call_graph(self, ssa=None):
-        cost = {Toffoli(): self.n_ctrls - 1}
+        cost: Dict[Bloq, int] = {Toffoli(): self.n_ctrls - 1}
         n_neg = sum(int(cv == 0) for cv in self.concrete_cvs)
         if n_neg:
             cost[XGate()] = 2 * n_neg
diff --git a/qualtran/bloqs/mcmt/approx_multi_toffoli_test.py b/qualtran/bloqs/mcmt/approx_multi_toffoli_test.py
index 690db1b0..91bf15be 100644
--- a/qualtran/bloqs/mcmt/approx_multi_toffoli_test.py
+++ b/qualtran/bloqs/mcmt/approx_multi_toffoli_test.py
@@ -36,6 +36,6 @@ def test_approx_multi_toffoli_classical_randomized_notebook_example():
     sample_strings = tuple(random_sample_string(m) for _ in range(k))
     bloq = ApproxMultiToffoli(n=n, k=k, sample_strings=sample_strings)
 
-    _, _, target = bloq.call_classically(ctrl=int_to_bits(x_int, m))
+    _, _, target = bloq.call_classically(ctrl=int_to_bits(x_int, m))  # type: ignore
 
     assert target == int(x_int == all_ones)

```

mpharrigan commented:
We use a linter to statically detect other issues prior to runtime. You can run it with `check/pylint`. It can take a minute or two to run. The errors are somewhat self-explanatory but let me know if you have any questions

mpharrigan commented:
Finally, we use an automated code formatter. When the other checks are good to go, run `check/format-incremental --apply` to automatically format the code in the PR

chenyizhang2000 committed:
Generate the reference notebook and include it in the bloqs docs
table of contents. Address mypy typing issues from review.

chenyizhang2000 commented:
I've fixed the issues from running `check/mypy`;

`check/pylint` gives me the following warning message: Missing or wrong copyright notice (wrong-or-nonexistent-copyright-notice);

I ran `check/format-incremental --apply` and it reformatted 2 files: `approx_multi_toffoli_test.py` and `approx_multi_toffoli.py`

chenyizhang2000 commented:
Hi Matt, hope you’re doing well! I think I’ve addressed the previous comments and the CI checks seem to be passing now. What would be the next step from my side?

mhucka commented:
@chenyizhang2000 Just circling around to different PRs and trying to do a bit of housekeeping. For planning purposes, could you let us know what is the status of this PR?

chenyizhang2000 commented:
Thanks for checking in! This is still active. I believe I’ve addressed Matt’s previous comments and have been waiting for a follow-up review. I noticed GitHub still shows two failed checks from June; I’m happy to investigate or rerun them if needed.

mhucka commented:
> Thanks for checking in! This is still active. I believe I’ve addressed Matt’s previous comments and have been waiting for a follow-up review. I noticed GitHub still shows two failed checks from June; I’m happy to investigate or rerun them if needed.

Thank you. Could you update the branch? That will give us a better picture of the CI test results.

<div align="center">
<img width="80%" alt="image" src="https://github.com/user-attachments/assets/f953663c-97fd-4246-a3e0-c5e7d94b20a9" />
</div>

chenyizhang2000 commented:
Thanks! I’ve updated the branch by merging the current upstream/main into multi_qubit_toffoli_draft and pushed the result.
